internal/idle/idle_e2e_test.go

```diff
@@ -544,3 +548,77 @@ func (s) TestChannelIdleness_Connect(t *testing.T) {
 	// Verify that the ClientConn moves back to READY.
 	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
 }
+
+// runFunc runs f repeatedly until the context expires.
+func runFunc(ctx context.Context, f func()) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(10 * time.Millisecond):
+			f()
+		}
+	}
+}
+
+// Tests the scenario where there are concurrent calls to exit and enter idle
+// mode on the ClientConn. Verifies that there is no race under this scenario.
+func (s) TestChannelIdleness_RaceBetweenEnterAndExitIdleMode(t *testing.T) {
+	// Start a test backend and set the bootstrap state of the resolver to
+	// include this address. This will ensure that when the resolver is
+	// restarted when exiting idle, it will push the same address to grpc again.
+	r := manual.NewBuilderWithScheme("whatever")
+	backend := stubserver.StartTestService(t, nil)
+	defer backend.Stop()
+	r.InitialState(resolver.State{Addresses: []resolver.Address{{Addr: backend.Address}}})
+
+	// Create a ClientConn with a long idle_timeout. We will explicitly trigger
+	// entering and exiting IDLE mode from the test.
+	dopts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithResolvers(r),
+		grpc.WithIdleTimeout(30 * time.Minute),
+		grpc.WithDefaultServiceConfig(`{"loadBalancingConfig": [{"pick_first":{}}]}`),
+	}
+	cc, err := grpc.Dial(r.Scheme()+":///test.server", dopts...)
+	if err != nil {
+		t.Fatalf("grpc.Dial() failed: %v", err)
+	}
+	defer cc.Close()
+
+	enterIdle := internal.EnterIdleModeForTesting.(func(*grpc.ClientConn) error)
+	enterIdleFunc := func() {
+		if err := enterIdle(cc); err != nil {
+			t.Errorf("Failed to enter idle mode: %v", err)
+		}
+	}
+	exitIdle := internal.ExitIdleModeForTesting.(func(*grpc.ClientConn) error)
+	exitIdleFunc := func() {
+		if err := exitIdle(cc); err != nil {
+			t.Errorf("Failed to exit idle mode: %v", err)
+		}
+	}
+	// Spawn goroutines that call methods on the ClientConn to enter and exit
+	// idle mode concurrently for one second.
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	var wg sync.WaitGroup
+	wg.Add(4)
+	go func() {
+		runFunc(ctx, enterIdleFunc)
+		wg.Done()
+	}()
```

Comment:
Maybe a `for` loop around this instead of repeating?  Then you can use a `const` for `4`.

Reply:
I was able to save only a couple of lines because I need two `for` loops, one for the `enterIdleFunc` and one for the `exitIdleFunc`. So, decided to leave it as is.

Reply:
You weren't code golfing hard enough, then!:

```go
for _, f := range []func(){enterIdleFunc, enterIdleFunc, exitIdleFunc, exitIdleFunc} {
	go func() {
		runFunc(ctx, f)
		wg.Done
	}()
}
```